rebasebot/bot.py

```diff
@@ -105,7 +105,31 @@ def _needs_rebase(gitwd, source, dest):
     return True
 
 
-def _do_rebase(gitwd, source, dest):
+def _is_pr_merged(pr_number, source_repo):
+    logging.info("Checking that PR %s has been merged", pr_number)
+    gh_pr = source_repo.pull_request(pr_number)
+    return gh_pr.is_merged()
+
+
+def _add_to_rebase(commit_message, source_repo):
+    if commit_message.startswith("UPSTREAM: "):
+        commit_message = commit_message.removeprefix("UPSTREAM: ")
+        commit_tag = commit_message.split(":", 1)[0]
+        if commit_tag == "<drop>":
+            return False
+
+        if commit_tag == "<carry>":
+            return True
+
+        if commit_tag.isnumeric():
+            return not _is_pr_merged(int(commit_tag), source_repo)
+
+        raise Exception(f"Unknown commit message tag: {commit_tag}")
+
+    return True
```

Comment:
So if we find a commit that doesn't meet the pattern, we keep it without scrutiny? Should we not drop it since it's not an `UPSTREAM` commit? The rest of the logic should include any commits from the upstream base branch since we are running a cherry-pick? (I'm thinking about what happened with the autoscaler rebase)

Reply:
I thought about it... The main issue is that other teams rarely use UPSTREAM tags. For instance, https://github.com/openshift/aws-ebs-csi-driver/commit/32801fa66917a0e00d16271d64b0f4e4e8f199e9 and https://github.com/openshift/cloud-provider-openstack/commit/b4e0a642c05e375902cfb79ee6a53157eda93548. If we start dropping untagged commits, it may break their repos.

The solution I came up with is to add a policy option to the bot (like `--tag-policy`).
- `--tag-policy=none` will take all commits into the rebase PR, regardless of their tags, even including UPSTREAM: <drop>
- `--tag-policy=soft` is what I did here. If the commit has `UPSTREAM: something` it will be taken into account, otherwise we keep it.
- `--tag-policy=strict` is similar to the previous one, but it discards commits without "UPSTREAM:".

Reply:
I like the idea of a tag policy flag, that seems very elegant! I would personally want strict on all of our repos but others may choose what they like. Do you have time to add a flag?

Reply:
done!